Sources/ETCD/ETCDClient.swift

```diff
@@ -120,5 +123,129 @@ public final class EtcdClient: @unchecked Sendable {
+        let watchAsyncSequence = WatchAsyncSequence(grpcAsyncSequence: watchClient.watch(request))
+        return try await operation(watchAsyncSequence)
+    }
+    
+    /// Watches a specified key in the ETCD server.
+    ///
+    /// - Parameters:
+    ///   - key: The key for which the value is watched. Parameter is of type String.
+    ///   - operation: The operation to be run on the WatchAsyncSequence for the key.
+    public func watch<Result>(_ key: String, _ operation: (WatchAsyncSequence) async throws -> Result) async throws -> Result {
+        try await self.watch(key.utf8, operation)
+    }
+}
+
+public enum EventType {
+    case put
+    case delete
+
+    init(protoType: Int) {
+        switch protoType {
+        case 0:
+            self = .put
+        case 1:
+            self = .delete
+        default:
+            fatalError("Unsupported event type")
+        }
+    }
+
+    var protoType: Int {
+        switch self {
+        case .put:
+            return 0
+        case .delete:
+            return 1
+        }
+    }
+}
+
+public struct KeyValue {
+    public var key: Data
+    public var createRevision: Int64
+    public var modRevision: Int64
+    public var version: Int64
+    public var value: Data
+    public var lease: Int64
+
+    init(protoKeyValue: Etcdserverpb_KeyValue) {
+        self.key = protoKeyValue.key
+        self.createRevision = protoKeyValue.createRevision
+        self.modRevision = protoKeyValue.modRevision
+        self.version = protoKeyValue.version
+        self.value = protoKeyValue.value
+        self.lease = protoKeyValue.lease
+    }
+}
 
+
+public struct WatchEvent {
+    public var kv: KeyValue
+    public var prevKV: KeyValue?
+
+    init(protoEvent: Etcdserverpb_Event) {
+        self.kv = KeyValue(protoKeyValue: protoEvent.kv)
+        if let protoPrevKV = protoEvent.hasPrevKv ? protoEvent.prevKv : nil {
+            self.prevKV = KeyValue(protoKeyValue: protoPrevKV)
+        } else {
+            self.prevKV = nil
+        }
+    }
+}
+
+
+public struct WatchAsyncSequence: AsyncSequence {
+    public typealias Element = [WatchEvent]
+
+    private let grpcAsyncSequence: GRPCAsyncResponseStream<Etcdserverpb_WatchResponse>
+
+    init(grpcAsyncSequence: GRPCAsyncResponseStream<Etcdserverpb_WatchResponse>) {
+        self.grpcAsyncSequence = grpcAsyncSequence
+    }
+
+    public func makeAsyncIterator() -> AsyncIterator {
+        .init(grpcIterator: self.grpcAsyncSequence.makeAsyncIterator())
+    }
+
+    public struct AsyncIterator: AsyncIteratorProtocol {
+        private var grpcIterator: GRPCAsyncResponseStream<Etcdserverpb_WatchResponse>.AsyncIterator?
+
+        fileprivate init(grpcIterator: GRPCAsyncResponseStream<Etcdserverpb_WatchResponse>.AsyncIterator) {
+            self.grpcIterator = grpcIterator
+        }
+
+        public mutating func next() async throws -> Element? {
+            while true {
```

Comment:
Instead of doing while `true` we could be doing `while let response = try await self.grpcIterator?.next()` and then change the inners of the loop accordingly

Reply:
Updated!